### PR TITLE
Add a UDF pair for simple random sampling with replacement (SRSWR).

### PR DESCRIPTION
**Pull request by mengxr**
Add a UDF pair that implements a scalable algorithm for simple random sampling with replacement. commons-math3-3.2 is used to generate binomial random variables.

### CODE REVIEW

**Comment by mengxr** [comment]
Please check the javadoc for usage.

File: src/java/datafu/pig/sampling/SimpleRandomSampleWithReplacementElect.java [hunk 1]
```diff
@@ -0,0 +1,182 @@
+package datafu.pig.sampling;
```
Comment (matthayes) [inline src/java/datafu/pig/sampling/SimpleRandomSampleWithReplacementElect.java:17]: Needs header

File: src/java/datafu/pig/sampling/SimpleRandomSampleWithReplacementVote.java [hunk 2]
```diff
@@ -0,0 +1,227 @@
+package datafu.pig.sampling;
```
Comment (matthayes) [inline src/java/datafu/pig/sampling/SimpleRandomSampleWithReplacementVote.java:17]: Needs header

File: test/pig/datafu/test/pig/sampling/SimpleRandomSampleWithReplacementTest.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
+package datafu.test.pig.sampling;
```
Comment (matthayes) [inline test/pig/datafu/test/pig/sampling/SimpleRandomSampleWithReplacementTest.java:17]: Needs header

File: ivy.xml [hunk 4]
```diff
@@ -14,16 +14,16 @@
         <dependency org="org.apache.pig" name="pigunit" rev="${pig.version}" conf="common->default"/>
         
         <!-- use the right avro version which pig 0.10.0 was built against -->
-        <dependency org="org.apache.avro" name="avro" rev="${avro.version}" conf="common->default"/>
-        
+        <dependency org="org.apache.avro" name="avro" rev="${avro.version}" conf="common->default"/>        
         <dependency org="it.unimi.dsi" name="fastutil" rev="${fastutil.version}" conf="common->default"/>
         <dependency org="joda-time" name="joda-time" rev="${joda-time.version}" conf="common->default"/>
         <dependency org="org.apache.commons" name="commons-math" rev="${commons-math.version}" conf="common->default"/>
+        <dependency org="org.apache.commons" name="commons-math3" rev="${commons-math3.version}" conf="common->default"/>
```
Comment (matthayes) [inline ivy.xml]: You should be able to remove this.  I have updated commons-math to 2.2.

File: ivy/libraries.properties [hunk 5]
```diff
@@ -1,6 +1,7 @@
 antlr.version=3.2
 avro.version=1.5.3
 commons-math.version=2.1
+commons-math3.version=3.2
```
Comment (matthayes) [inline ivy/libraries.properties]: can remove

File: src/java/datafu/pig/sampling/SimpleRandomSampleWithReplacementVote.java [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+
+    DataBag items = (DataBag) tuple.get(0);
+    int sampleSize = ((Number) tuple.get(1)).intValue();
+    long count = ((Number) tuple.get(2)).longValue();
+
+    /*
+     * The following threshold is to guarantee that each output position contains at least
+     * one candidate with high probability.
+     */
+    double threshold = 1.0d - Math.exp(Math.log(FAILURE_RATE / sampleSize) / count);
```
Comment (matthayes) [inline src/java/datafu/pig/sampling/SimpleRandomSampleWithReplacementVote.java:216]: Can you explain this threshold computation?  is it in the paper?  I couldn't find it.

Comment (mengxr) [inline src/java/datafu/pig/sampling/SimpleRandomSampleWithReplacementVote.java:216]: It is not in the paper but it is a simple extension of that paper. I explained the algorithm briefly in the last paragraph of the javadoc. Basically, for each output position, we do a random sort and select the smallest one. Let s be the sample size and n be the population size. For any q > 0, the probability of all random keys being greater than q is (1-q)^n. Thus, if we throw away items with associated keys greater than q, with probability at least 1 - s (1-q)^n, we still have at least one candidate for each position. Let the failure rate delta = s (1-q)^n and solve for q, we get q = 1 - exp(log(delta/s)/n).

Comment (mengxr) [inline src/java/datafu/pig/sampling/SimpleRandomSampleWithReplacementVote.java:216]: Added more about the algorithm in the javadoc. Please check the latest commit.

**Comment by matthayes** [comment]
Thanks!  The doc changes look great.  Also wanted to say your docs in general are excellent :)  Thanks for all the details.

**Pull request merged**